moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java

```diff
@@ -985,4 +985,98 @@ JsonReader newReader(String json) throws IOException {
     reader.readJsonValue();
     assertThat(reader.hasNext()).isFalse();
   }
+
+  @Test public void basicPeekJson() throws IOException {
+    JsonReader reader = newReader("{\"a\":12,\"b\":[34,56],\"c\":78}");
+    assumeTrue(reader instanceof JsonValueReader); // Not implemented for JsonUtf8Reader yet!
+    reader.beginObject();
+    assertThat(reader.nextName()).isEqualTo("a");
+    assertThat(reader.nextInt()).isEqualTo(12);
+    assertThat(reader.nextName()).isEqualTo("b");
+    reader.beginArray();
+    assertThat(reader.nextInt()).isEqualTo(34);
+
+    // Peek.
+    JsonReader peekReader = reader.peekJson();
+    assertThat(peekReader.nextInt()).isEqualTo(56);
+    peekReader.endArray();
+    assertThat(peekReader.nextName()).isEqualTo("c");
+    assertThat(peekReader.nextInt()).isEqualTo(78);
+    peekReader.endObject();
+    assertThat(peekReader.peek()).isEqualTo(JsonReader.Token.END_DOCUMENT);
+
+    // Read again.
+    assertThat(reader.nextInt()).isEqualTo(56);
+    reader.endArray();
+    assertThat(reader.nextName()).isEqualTo("c");
+    assertThat(reader.nextInt()).isEqualTo(78);
+    reader.endObject();
+    assertThat(reader.peek()).isEqualTo(JsonReader.Token.END_DOCUMENT);
+  }
+
+  @Test public void peekJsonReader() throws IOException {
+    JsonReader reader = newReader("[12,34,{\"a\":56,\"b\":78},90]");
+    assumeTrue(reader instanceof JsonValueReader); // Not implemented for JsonUtf8Reader yet!
+    for (int i = 0; i < 12; i++) {
+      readPeek12Steps(reader.peekJson(), i, 12);
+      readPeek12Steps(reader, i, i + 1);
+    }
+  }
+
+  /**
+   * Read a fragment of {@code reader}. This assumes the fixed document defined in {@link
+   * #peekJsonReader} and reads a range of it on each call.
+   */
+  private void readPeek12Steps(JsonReader reader, int from, int until) throws IOException {
```

Comment:
is this simply `if (from == until) return;`?

Reply:
Not exactly.

Reply:
forgot switch statements today. 👍